### PR TITLE
Gconstruct hard negative

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
Support processing edge hard negatives in gconstruct.

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/transform.py [hunk 1]
```diff
@@ -935,6 +935,143 @@ def call(self, feats):
...
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
+            feats = feats.astype(str)
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1018]: what is `str`?

File: python/graphstorm/gconstruct/transform.py [hunk 2]
```diff
@@ -935,6 +935,143 @@ def call(self, feats):
...
+            print(feats.dtype.type)
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1019]: remove print

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -935,6 +935,143 @@ def call(self, feats):
...
+                dim_size = len(feat.split(self._separator))
+                max_dim = dim_size if dim_size > max_dim else max_dim
+        return {self.feat_name: max_dim}
+
+    def update_info(self, info):
+        """ Store global information for the second phase data processing
+
+        Parameters
+        ----------
+        info:
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1031]: what is the type of `info`?

File: python/graphstorm/gconstruct/transform.py [hunk 4]
```diff
@@ -935,6 +935,143 @@ def call(self, feats):
...
+            Information to be collected
+        """
+        max_dim = max(info)
+        self._max_dim = max_dim
+
+    def call(self, feats):
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1037]: what is `feats`

File: python/graphstorm/gconstruct/transform.py [hunk 5]
```diff
@@ -935,6 +935,143 @@ def call(self, feats):
                 f"The feature {self.feat_name} has to be integers or floats."
         return {self.feat_name: feats}
 
+class HardEdgeNegativeTransform(TwoPhaseFeatTransform):
+    """ Translate input data into node ids for hard negative stored as edge features
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:939]: do i understand correctly that users prepare the hard negative samples and stores them as edge features? this class interprets the edge features and loads them as hard negative edges.

Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:939]: are these negative edges stored in the DGL graph as edge features? the data loader needs to understand the edge features.

File: python/graphstorm/gconstruct/transform.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1018]: We are converting feats to string numpy array. (When loading data from Parquet, sometimes, the numpy array dtype is `object`.

File: python/graphstorm/gconstruct/transform.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1019]: removed

File: python/graphstorm/gconstruct/transform.py (hunk 3 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1031]: statistics return by pre_process.

File: python/graphstorm/gconstruct/transform.py (hunk 4 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1037]: Updated

File: python/graphstorm/gconstruct/transform.py [hunk 6]
```diff
@@ -935,6 +910,149 @@ def call(self, feats):
...
+        Return:
+            dict: {feature_name: feats_statistics}
+        """
+        assert isinstance(feats, (np.ndarray, ExtMemArrayWrapper)), \
+            "Feature of HardEdgeNegativeTransform must be numpy array or ExtMemArray"
+
+        if self._separator is None:
+            max_dim = feats.shape[1]
+        else:
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:987]: should you check the dtype here if the input has to be strings?

File: python/graphstorm/gconstruct/transform.py [hunk 7]
```diff
@@ -935,6 +910,149 @@ def call(self, feats):
...
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:995]: is it better to assert here? if the input isn't string, even if we convert the dtype, does it still work?

File: python/graphstorm/gconstruct/transform.py [hunk 8]
```diff
@@ -935,6 +910,149 @@ def call(self, feats):
...
+        np.array
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        nid_map = self._nid_map[self._target_ntype]
+
+        neg_ids = np.full((len(feats), self._max_dim), -1, dtype=np.int64)
+        if self._separator is None:
+            for i, raw_ids in enumerate(feats):
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1026]: it's kind of confusing here. a row of `feats` is a list of ids, separated by some delimitor?

File: python/graphstorm/gconstruct/transform.py [hunk 9]
```diff
@@ -935,6 +910,149 @@ def call(self, feats):
...
+                if raw_ids is None:
+                    continue
+                # raw_ids is a numpy array
+                # handle empty strings in raw_ids
+                raw_ids = raw_ids.astype(str)
+                nid_idx = np.nonzero(raw_ids != "")[0]
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1032]: what kind of input do you expect? the negative edges should be prepared by users. do you expect there are many different corner cases?

Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:1032]: or maybe you can give an example of how the negative edges are stored?

File: python/graphstorm/gconstruct/transform.py [hunk 10]
```diff
@@ -935,6 +910,149 @@ def call(self, feats):
...
+                nids, _ = nid_map.map_id(raw_ids[nid_idx].astype(
+                    nid_map.map_key_dtype))
+                neg_ids[i][nid_idx] = nids
+        else:
+            for i, feat in enumerate(feats):
+                if feat is None:
+                    continue
+                raw_ids = np.array(feat.split(self._separator))
+                nids, _ = nid_map.map_id(raw_ids.astype(nid_map.map_key_dtype))
+                neg_ids[i][:nids.shape[0]] = nids
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1042]: do you expect that different edges have different numbers of negative edges?

File: python/graphstorm/gconstruct/utils.py [hunk 11]
```diff
@@ -979,3 +999,77 @@ def partition_graph(g, node_data, edge_data, graph_name, num_partitions, output_
...
+            hard_edge_neg_feats[edge_type] = {neg_ntype: [hard_edge_neg_op.feat_name]}
+        else:
+            if neg_ntype in hard_edge_neg_feats[edge_type]:
+                hard_edge_neg_feats[edge_type][neg_ntype].append(hard_edge_neg_op.feat_name)
+            else:
+                hard_edge_neg_feats[edge_type][neg_ntype] = [hard_edge_neg_op.feat_name]
+
+    return hard_edge_neg_feats
+
+def shuffle_hard_nids(data_path, num_parts, hard_edge_neg_feats):
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/utils.py:1026]: do you mean shuffle or map?

File: python/graphstorm/gconstruct/transform.py (hunk 10 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1042]: It is possible.

File: python/graphstorm/gconstruct/utils.py (hunk 11 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:1026]: Shuffle the nids according to graph partition.

File: python/graphstorm/gconstruct/transform.py (hunk 9 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1064]: I have added some examples

File: docs/source/advanced/advanced-usages.rst [hunk 12]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
 
 With DGL 1.0.4, ``fast_localuniform`` dataloader can speedup 2.4X over ``localuniform`` dataloader on training a 2 layer RGCN on MAG dataset on four g5.48x instances.
 
+Hard Negative sampling in Link Prediction Training
+-------------------------------------
+GraphStorm provides support for users to define hard negative nodes for a positive edge during Link Prediction Training.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:27]:
```suggestion
GraphStorm provides support for users to define hard negative edges for a positive edge during Link Prediction Training.
```

File: docs/source/advanced/advanced-usages.rst [hunk 13]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+Currently, hard negative nodes can only be destination nodes.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:28]:
```suggestion
Currently, hard negative edges are constructed by replacing the destination nodes of an edge with randomly sampled nodes.
```

File: docs/source/advanced/advanced-usages.rst [hunk 14]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+For example, given an edge (``src_pos``, ``dst_pos``) and hard negative nodes ``hard_0`` and ``hand_1``, GraphStorm will construct two hard negative edges, i.e., (``src_pos``, ``hard_0``) and (``src_pos``, ``hand_1``).
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:29]: i think it's better to not use hard negative nodes. I don't think the concept is well defined.

File: docs/source/advanced/advanced-usages.rst [hunk 15]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+      - src_type,rel_type0,dst_type:5
+      - src_type,rel_type1,dst_type:10
+
+Hard negative sampling can be used together with any link prediction negative sampler, such as ``uniform``, ``joint``, ``inbatch_joint``, etc.
+By default, GraphStorm will sample hard negatives first to fulfill the requirement of ``num_hard_negatives`` and then sample random negatives to fulfill the requirement of ``num_negative_edges``.
+In general, GraphStorm covers following cases:
+
+- ``num_hard_negatives`` is larger or equal to ``num_negative_edges``. GraphStorm will only sample hard negative nodes.
+- There are enough hard negatives for a positive edge. GraphStorm will randomly sample ``num_hard_negatives`` hard negative nodes from the hard negative set and then randomly sample ``num_negative_edges - num_hard_negatives`` negative nodes.
+- There is not enough hard negatives for a positive edge. GraphStorm will sample all the hard negatives first and then randomly sample negative nodes to fulfill the requirement of ``num_negative_edges``
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:56]:
```suggestion
- There is not enough hard negatives for a positive edge. GraphStorm will use all the hard negatives first and then randomly sample negative nodes to fulfill the requirement of ``num_negative_edges``
```

File: docs/source/advanced/advanced-usages.rst [hunk 16]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+
+** Preparing graph data for hard negative sampling **
+
+The gconstruct pipeline of GraphStorm provides support to load hard negative data from raw input.
+Hard destination negatives can be defined through ``edge_dst_hard_negative`` transformation.
+The ``feature_col`` field of ``edge_dst_hard_negative`` must stores the raw node ids of hard destination nodes.
+GraphStorm accepts two types of hard negative inputs:
+
+- **An array of strings** When the input format is ``Parquet`` or ``HDFS``, the ``feature_col`` can store string arrays. In this case, each row stores a string array representing the hard negative node ids of the corresponding edge. The ``feature_col`` will be a 2D string array, for example ``[["e0_hard_0", "e0_hard_1"],["e1_hard_0", "e1_hard_1"], ..., ["en_hard_0", "en_hard_1"]]``. It is required that each row has the same dimension size. In case when some edges do not have enough pre-defined hard negatives, GraphStorm allows users to use an empty string to fill the array. GraphStorm will automatically handle the case when some edges do not have enough hard negatives.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:65]: why do we allow some edges not to have enough pre-defined hard negative edges?

File: docs/source/advanced/advanced-usages.rst [hunk 17]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+
+- **A single string** The ``feature_col`` stores strings instead of string arrays. In this case, a ``separator`` must be provided to split the strings into node ids. The ``feature_col`` will be a 1D string list, for example ``["e0_hard_0;e0_hard_1", "e1_hard_0;e1_hard_1", ..., "en_hard_0;en_hard_1"]``. The string length, i.e., number of hard negatives, can vary from row to row. GraphStorm will automatically handle the case when some edges do not have enough hard negatives.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:67]: why do we want to handle different numbers of hard negative edges?

File: python/graphstorm/gconstruct/transform.py [hunk 18]
```diff
@@ -935,6 +910,185 @@ def call(self, feats):
...
+                dim_size = len(feat.split(self._separator))
+                max_dim = dim_size if dim_size > max_dim else max_dim
+        return {self.feat_name: max_dim}
+
+    def update_info(self, info):
+        max_dim = max(info)
+        self._max_dim = max_dim
+
+    def call(self, feats):
+        """ Generate hard negatives as features
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1008]:
```suggestion
        """ Parse hard negatives as features
```

File: python/graphstorm/gconstruct/transform.py (hunk 10 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:1078]: should we consider this an incorrect input?

Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:1078]: currently, if an edge doesn't have enough hard negative edges, it is filled with -1. we need to carefully handle -1 during the training. is it necessary to introduce the complexity?

File: python/graphstorm/gconstruct/transform.py (hunk 7 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:995]: We can not assert it. If we load data from parquet, the dtype will be np.object. Thus, we need to do the conversion. This is a hint for user that the data stored should be string.

File: python/graphstorm/gconstruct/transform.py (hunk 10 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:1078]: if there aren't enough hard negative edges, we can just duplicate them. the impact should be the same.

File: docs/source/advanced/advanced-usages.rst [hunk 19]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+    num_train_hard_negatives:
+      - src_type,rel_type0,dst_type:5
+      - src_type,rel_type1,dst_type:10
+
+Hard negative sampling can be used together with any link prediction negative sampler, such as ``uniform``, ``joint``, ``inbatch_joint``, etc.
+By default, GraphStorm will sample hard negatives first to fulfill the requirement of ``num_train_hard_negatives`` and then sample random negatives to fulfill the requirement of ``num_negative_edges``.
+In general, GraphStorm covers following cases:
+
+- ``num_train_hard_negatives`` is larger or equal to ``num_negative_edges``. GraphStorm will only sample hard negative nodes.
+- There are enough hard negatives for a positive edge. GraphStorm will randomly sample ``num_train_hard_negatives`` hard negative nodes from the hard negative set and then randomly sample ``num_negative_edges - num_train_hard_negatives`` negative nodes.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:55]: i don't understand this case. how is it different from the first case?

File: docs/source/advanced/advanced-usages.rst [hunk 20]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+
+** Preparing graph data for hard negative sampling **
+
+The gconstruct pipeline of GraphStorm provides support to load hard negative data from raw input.
+Hard destination negatives can be defined through ``edge_dst_hard_negative`` transformation.
+The ``feature_col`` field of ``edge_dst_hard_negative`` must stores the raw node ids of hard destination nodes.
+GraphStorm accepts two types of hard negative inputs:
+
+- **An array of strings** When the input format is ``Parquet`` or ``HDFS``, the ``feature_col`` can store string arrays. In this case, each row stores a string array representing the hard negative node ids of the corresponding edge. The ``feature_col`` will be a 2D string array, for example ``[["e0_hard_0", "e0_hard_1"],["e1_hard_0", "e1_hard_1"], ..., ["en_hard_0", "en_hard_1"]]``. It is required that each row has the same dimension size. In case when some edges do not have enough pre-defined hard negatives, GraphStorm allows users to use an empty string to fill the array. GraphStorm will automatically handle the case when some edges do not have enough hard negatives.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:65]: are you sure HDF5 works with strings? Qi said it uses weird encoding.

File: python/graphstorm/gconstruct/transform.py (hunk 7 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:995]: if so, should we print warning? it seems that it's a normal behavior that dtype is not string.

File: python/graphstorm/gconstruct/transform.py [hunk 21]
```diff
@@ -935,6 +910,185 @@ def call(self, feats):
...
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        nid_map = self._nid_map[self._target_ntype]
+
+        # It is possible that some edges do not
+        # have enough pre-defined hard negatives.
+        # In certain cases, GraphStorm will fill the
+        # un-provided hard negatives with -1s.
+        neg_ids = np.full((len(feats), self._max_dim), -1, dtype=np.int64)
+        if self._separator is None:
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:1053]: did you test this on HDF5?

File: python/graphstorm/gconstruct/transform.py (hunk 7 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:995]: make sense

File: docs/source/advanced/advanced-usages.rst (hunk 19 above)
Comment (classicsong) [reply docs/source/advanced/advanced-usages.rst:55]: There are different cases:
First case, num_train_hard_negatives > num_negative_edges. That means all negatives will be sampled from hard negatives.
Second case, num_train_hard_negatives < num_negative_edges. Part of negatives are from hard negatives and the rest from random negative sampler.

Comment (classicsong) [reply docs/source/advanced/advanced-usages.rst:55]: Updated the description.

File: docs/source/advanced/advanced-usages.rst (hunk 20 above)
Comment (classicsong) [reply docs/source/advanced/advanced-usages.rst:65]: Let's exclude HDF5

File: python/graphstorm/gconstruct/transform.py (hunk 21 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1053]: No. Let me add a test.

Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1053]: Encoding string hdf5 is a little bit complex, maybe we can just skip supporting hdf5.

File: docs/source/advanced/advanced-usages.rst [hunk 22]
```diff
@@ -22,6 +22,52 @@ To be more specific, these dataloaders will do neighbor sampling regardless of a
...
+
+- **An array of strings** When the input format is ``Parquet``, the ``feature_col`` can store string arrays. In this case, each row stores a string array representing the hard negative node ids of the corresponding edge. The ``feature_col`` will be a 2D string array, for example ``[["e0_hard_0", "e0_hard_1"],["e1_hard_0", "e1_hard_1"], ..., ["en_hard_0", "en_hard_1"]]``. It is required that each row has the same dimension size. In case when some edges do not have enough pre-defined hard negatives, GraphStorm allows users to use an empty string to fill the array. GraphStorm will automatically handle the case when some edges do not have enough hard negatives.
+
+- **A single string** The ``feature_col`` stores strings instead of string arrays. (When the input format is ``Parquet`` or ``CSV``) In this case, a ``separator`` must be provided to split the strings into node ids. The ``feature_col`` will be a 1D string list, for example ``["e0_hard_0;e0_hard_1", "e1_hard_0;e1_hard_1", ..., "en_hard_0;en_hard_1"]``. The string length, i.e., number of hard negatives, can vary from row to row. GraphStorm will automatically handle the case when some edges do not have enough hard negatives.
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:67]: another format is an array of integers.

**Review by zheng-da — APPROVED** [review]

**Pull request merged**